src/borrow.rs

```diff
@@ -493,14 +646,59 @@ fn base_address<T, D>(array: &PyArray<T, D>) -> usize {
     }
 }
 
+fn data_range<T, D>(array: &PyArray<T, D>) -> Range<usize>
+where
+    T: Element,
+    D: Dimension,
+{
+    let data = unsafe { (*array.as_array_ptr()).data };
+
+    let mut start = 0;
+    let mut end = 0;
+
+    for (&dim, &stride) in array.shape().iter().zip(array.strides()) {
+        let offset = (dim as isize) * stride;
+
+        if offset >= 0 {
+            end += offset;
+        } else {
+            start += offset;
+        }
+    }
```

Comment:
Is this correct?
For example, for an f64 array with shape `(2, 3, 4)` (thus of which stride is `(96, 32, 8)`), I think 96 * 2 points out the next pointer to this array (i.e., the nearest pointer from the last element).

Reply:
Indeed, I think it is _not_ correct. I was going of the usual inclusive lower and exclusive upper bound but this does not work like this of multiple dimensions. I think I will have to use `let offset = ((dim -  1) as isize) * stride)`  and then add one to `end` at the end... (And my test cases are thereby bogus as well. |-\)

Thanks for catching this!

Reply:
I fixed a push as a separate commit.

I switch to an inclusive instead of an exclusive range to avoid having to make `data_range` depend on item size and adjusted the check accordingly. I also replaced `std::ops::Range` by `(usize, usize)` as the former is documented to be exclusive and `std::ops::RangeInclusive` is not helpful as it contains an extra "exhausted" field which we do not need (as it is an iterator).

Could you have another look? Thank you!